src/adapter/src/client.rs

```diff
@@ -397,13 +400,79 @@ impl SessionClient {
     pub async fn simple_execute(
         &mut self,
         stmts: &str,
+        params: &[(&str, &str)],
     ) -> Result<SimpleExecuteResponse, AdapterError> {
+        // To support parameters we do an AST rewrites taking `$x` to
+        // `CAST($params[x-1].0 AS $params[x-1].1)`
+        fn inline_params(
+            params: &[(&str, &str)],
+            stmt: &mut mz_sql::ast::Statement<Raw>,
+        ) -> Result<(), anyhow::Error> {
+            struct ParamInliner<'a> {
+                params: &'a [(&'a str, &'a str)],
+                error: Option<anyhow::Error>,
+            }
+
+            impl<'ast, 'a> VisitMut<'ast, Raw> for ParamInliner<'a> {
+                fn visit_expr_mut(&mut self, node: &'ast mut Expr<Raw>) {
+                    if let Expr::Parameter(i) = node {
+                        let i = *i;
+                        // 1-indexed params make this awk
+                        let vals = match i {
+                            i if i == 0 => None,
+                            i => self.params.get(i - 1),
+                        };
+
+                        let (val, typ) = match vals {
+                            Some(v) => v,
+                            None => {
+                                self.error = Some(anyhow!("there is no parameter ${}", i));
+                                return;
+                            }
+                        };
+
+                        match mz_sql_parser::parser::parse_expr(&format!(
+                            "CAST ('{}' AS {})",
+                            display::escape_single_quote_string(val),
+                            typ
```

Comment:
Using the `typ` as passed by the user without validating it is a means for SQL injection if we use string formatting.

Reply:
It is validated; the second argument can only be parsed as a data type. Do you have any suggestions for a means by which this could be injected into?

Reply:
I had in mind something like `int) ...`. Can we explicitly parse this as a type and then inject the `to_string` of that here?